<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>the Nanopore `Sequencing Summary` file • floundeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="the Nanopore `Sequencing Summary` file">
<meta property="og:description" content="floundeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">floundeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_fasta_files.html">Fasta format sequence files</a>
    </li>
    <li>
      <a href="../articles/02_fastq_files.html">Fastq format sequence files</a>
    </li>
    <li>
      <a href="../articles/03_fast5_files.html">Fast5 format sequence files</a>
    </li>
    <li>
      <a href="../articles/04_seqsum_files.html">the Nanopore `Sequencing Summary` file</a>
    </li>
    <li>
      <a href="../articles/05_blast_files.html">Blast result parsing</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="04_seqsum_files_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="04_seqsum_files_files/kePrint-0.0.1/kePrint.js"></script><link href="04_seqsum_files_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>the Nanopore <code>Sequencing Summary</code> file</h1>
            
      
      
      <div class="hidden name"><code>04_seqsum_files.Rmd</code></div>

    </div>

    
    
<div id="preamble" class="section level2">
<h2 class="hasAnchor">
<a href="#preamble" class="anchor"></a>Preamble</h2>
<p>This <code>sequencing_summary_review</code> vignette re-introduces the <code>BasicQC</code>-type tutorial that was previously available through the <a href="https://github.com/nanoporetech" class="uri">https://github.com/nanoporetech</a> website.</p>
<p>The workflow provided is implemented through the R TidyVerse and is intended to provide a lightweight set of components that can be mixed and matched for the exploration of the <code>sequencing_summary</code> file that is produced during Guppy-based base-calling.</p>
<p>This report is largely feature-complete with the earlier <code>BasicQC</code> tutorial but has been implemented with a different set of rules and expectations. A couple of figures previously reported have been deprecated for aesthetic reasons. Feedback is welcomed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">floundeR</span><span class="op">)</span>
<span class="co">#&gt; floundeR v0.0.5</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/">kableExtra</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="using-a-real-world-sequencing_summary-file-" class="section level2">
<h2 class="hasAnchor">
<a href="#using-a-real-world-sequencing_summary-file-" class="anchor"></a>Using a real-world sequencing_summary file.</h2>
<p>The <code>floundeR</code> R package provides an example <em>sequencing_summary</em> file. This is 10,000 sequence reads long and provides insight into what is possible with the software. For this demonstration we’ll have a look at something a little more impressive and illustrative of what the software is intended to do.</p>
<p><a href="https://registry.opendata.aws/ont-open-data/">s3://ont-open-data</a> is a data repository hosted by Amazon. The human genome, GM24385 / HG002, has been sequenced on a PromethION and deposited at the registry <a href="https://nanoporetech.github.io/ont-open-datasets/gm24385_2020.11/">read the blog post</a>. We will download the dataset to demonstrate what can be done.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sequencing_summary</span> <span class="op">&lt;-</span> <span class="st">"sequencing_summary_PAG07165_2dfda515.txt"</span>

<span class="co"># don't perform the AWS download if the file has already been grabbed ...</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">sequencing_summary</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">aws.s3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aws.s3/man/get_object.html">save_object</a></span><span class="op">(</span>
  <span class="st">"/gm24385_2020.11/flowcells/20201026_1645_6B_PAG07165_d42912aa/sequencing_summary_PAG07165_2dfda515.txt"</span>,
  bucket<span class="op">=</span><span class="st">"s3://ont-open-data/"</span>, 
  region<span class="op">=</span><span class="st">"eu-west-1"</span>, 
  overwrite<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="define-input-files-and-create-the-sequencingsummary-r6-object" class="section level2">
<h2 class="hasAnchor">
<a href="#define-input-files-and-create-the-sequencingsummary-r6-object" class="anchor"></a>Define input files and create the <code>SequencingSummary</code> R6 object</h2>
<p>The <code>floundeR</code> package has been implemented largely as object oriented R through the usage of <code>R6 objects</code>. This means that instead of passing tables of data between methods that filter and visualise data facets we will be processing objects that contain their own methods and filters.</p>
<p>The first step in the process is to create out very first object; this is a <code>SequencingSummary</code> object and it is instantiated by calling its <code>initialize</code> constructor with the <code>SequencingSummary$new()</code> call. We will pass the <code>file.path</code> to our sequencing summary file.</p>
<p>Let’s also check that the object creation has worked by asking about the flowcell platform - this should return the expected sequencing platform.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># sequencing_summary &lt;- flnDr("sequencing_summary.txt.bz2")</span>
<span class="va">seqsum</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SequencingSummary.html">SequencingSummary</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">sequencing_summary</span><span class="op">)</span>

<span class="co"># check the flowcell platform - ensures that something valid has been created</span>
<span class="va">seqsum</span><span class="op">$</span><span class="va">flowcell</span><span class="op">$</span><span class="va">platform</span>
<span class="co">#&gt; Preparing channel count information</span>
<span class="co">#&gt; [1] "PromethION"</span></code></pre></div>
<div id="what-is-thismischief" class="section level3">
<h3 class="hasAnchor">
<a href="#what-is-thismischief" class="anchor"></a>What is this$mischief?</h3>
<p>As discussed in the section above, the <code>floundeR</code> objects have been wrapped as R6 objects. In the code <code>seqsum</code> is an object and the <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> is used to access stored variables, functions and active-bindings. It just happens that <code>seqsum$flowcell</code> returns a <code>Flowcell</code> object and this has a method called <code>platform</code> that returns the sequencing platform used.</p>
<p>While these classes make life simpler this can create some slightly scary code; within this package we have tried to ensure that the methods accessible through the R6 objects are also available through <code>magrittr::%&gt;%</code> pipes - the author of this package finds that this often makes for simpler code.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># let's see if things are really similar</span>
<span class="va">seqsum</span><span class="op">$</span><span class="va">flowcell</span>
<span class="co">#&gt; &lt;floundeR::Flowcell&gt;</span>

<span class="va">seqsum</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/to_flowcell.html">to_flowcell</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; &lt;floundeR::Flowcell&gt;</span></code></pre></div>
<p>The majority of <code>floundeR</code> classes provide a simple <code>$as_tibble()</code> method that returns a <code>tibble</code> of data that the class operates on. For the <code>SequencingSummary</code> object this is the whole imported file - we can have a check of the imported data using the following command</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seqsum</span><span class="op">$</span><span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>, 
             caption<span class="op">=</span><span class="st">"An excerpt of imported data as prepared by $as_tibble"</span>, 
             booktabs<span class="op">=</span><span class="cn">TRUE</span>, table.envir<span class="op">=</span><span class="st">'table*'</span>, linesep<span class="op">=</span><span class="st">""</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html">kable_styling</a></span><span class="op">(</span>latex_options<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"hold_position"</span>, font_size<span class="op">=</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table table">
<caption>
An excerpt of imported data as prepared by $as_tibble
</caption>
<thead><tr>
<th style="text-align:right;">
channel
</th>
<th style="text-align:right;">
start_time
</th>
<th style="text-align:right;">
duration
</th>
<th style="text-align:left;">
passes_filtering
</th>
<th style="text-align:right;">
sequence_length_template
</th>
<th style="text-align:right;">
mean_qscore_template
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
2269
</td>
<td style="text-align:right;">
63.87225
</td>
<td style="text-align:right;">
0.89850
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
383
</td>
<td style="text-align:right;">
13.339876
</td>
</tr>
<tr>
<td style="text-align:right;">
2757
</td>
<td style="text-align:right;">
63.62725
</td>
<td style="text-align:right;">
0.98925
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
267
</td>
<td style="text-align:right;">
3.216377
</td>
</tr>
<tr>
<td style="text-align:right;">
1818
</td>
<td style="text-align:right;">
63.63175
</td>
<td style="text-align:right;">
2.67975
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
1373
</td>
<td style="text-align:right;">
13.076194
</td>
</tr>
<tr>
<td style="text-align:right;">
1846
</td>
<td style="text-align:right;">
65.79525
</td>
<td style="text-align:right;">
1.08700
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
422
</td>
<td style="text-align:right;">
9.036358
</td>
</tr>
<tr>
<td style="text-align:right;">
1565
</td>
<td style="text-align:right;">
63.73250
</td>
<td style="text-align:right;">
3.51100
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
1701
</td>
<td style="text-align:right;">
12.828937
</td>
</tr>
<tr>
<td style="text-align:right;">
998
</td>
<td style="text-align:right;">
63.80150
</td>
<td style="text-align:right;">
2.95275
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
1283
</td>
<td style="text-align:right;">
11.641397
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="the-sequencingset-and-primitive-qc-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#the-sequencingset-and-primitive-qc-measures" class="anchor"></a>The <code>SequencingSet</code> and primitive QC measures</h2>
<p>In the previous section we created our <code>SequencingSummary</code> object. We can obtain a <code>SequencingSet</code> from the <code>SequencingSummary</code> object by calling its <code>$sequencingset</code> active binding; here we’ll use the <code>magrittr::%&gt;%</code> pipe for simplicity.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seqsum</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/to_sequencing_set.html">to_sequencing_set</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; &lt;floundeR::SequencingSet&gt;</span></code></pre></div>
<p>The <code>SequencingSet</code> in turn has a collection of methods that can be used to structure and visualise the data. The first that we’ll have a look at is the <code>$enumerate</code> method that returns an <code>Angenieux</code> object for data visualisation.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span>
  <span class="va">seqsum</span><span class="op">$</span><span class="va">sequencingset</span><span class="op">$</span><span class="va">enumerate</span><span class="op">$</span><span class="fu">to_file</span><span class="op">(</span><span class="st">"figure_2.png"</span><span class="op">)</span><span class="op">$</span><span class="fu">plot</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; saving plot as [png]</span></code></pre></div>
<p><img src="figure_2.png" width="75%"></p>
<p>There are a plethora of ways through which the <code>Angenieux</code> object can be used to style, colour and manipulate the graph - please do have a look at the methods documentation.</p>
<p>The <code>SequencingSet</code> object can also be used to access simple but primitive summary statistics such as <strong><code>mean</code></strong> sequence length, <strong><code>N50</code></strong> length etc</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seqsum</span><span class="op">$</span><span class="va">sequencingset</span><span class="op">$</span><span class="va">N50</span>
<span class="co">#&gt; [1] 29925</span>
<span class="va">seqsum</span><span class="op">$</span><span class="va">sequencingset</span><span class="op">$</span><span class="va">mean</span>
<span class="co">#&gt; [1] 21933.22</span></code></pre></div>
<div id="review-sequence-length-distributions" class="section level3">
<h3 class="hasAnchor">
<a href="#review-sequence-length-distributions" class="anchor"></a>Review sequence length distributions</h3>
<p>The distribution of sequence lengths is an important metric that is impacted by choice of library preparation, starting DNA isolation etc. A plot of length distributions is prepared from the same <code>SequencingSet</code> object that we reviewed in the previous section.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span>
  <span class="va">seqsum</span><span class="op">$</span><span class="va">sequencingset</span><span class="op">$</span><span class="fu">read_length_bins</span><span class="op">(</span>bins<span class="op">=</span><span class="fl">35</span>, outliers<span class="op">=</span><span class="fl">0.001</span><span class="op">)</span><span class="op">$</span>
    <span class="fu">to_file</span><span class="op">(</span><span class="st">"figure_3.png"</span><span class="op">)</span><span class="op">$</span>
    <span class="fu">plot</span><span class="op">(</span>style<span class="op">=</span><span class="st">"stacked"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; saving plot as [png]</span></code></pre></div>
<p><img src="figure_3.png" width="75%"></p>
</div>
<div id="review-quality-distributions" class="section level3">
<h3 class="hasAnchor">
<a href="#review-quality-distributions" class="anchor"></a>Review quality distributions</h3>
<p>The distribution of sequence lengths is an important metric that is impacted by choice of library preparation, starting DNA isolation etc. A plot of length distributions is prepared from the same <code>SequencingSet</code> object that we reviewed in the previous section.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span>
  <span class="va">seqsum</span><span class="op">$</span><span class="va">sequencingset</span><span class="op">$</span><span class="fu">quality_bins</span><span class="op">(</span>bins<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">$</span>
    <span class="fu">to_file</span><span class="op">(</span><span class="st">"figure_4.png"</span><span class="op">)</span><span class="op">$</span>
    <span class="fu">plot</span><span class="op">(</span>style<span class="op">=</span><span class="st">"stacked"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; saving plot as [png]</span></code></pre></div>
<p><img src="figure_4.png" width="75%"></p>
</div>
</div>
<div id="plot-flowcell-spatial-density" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-flowcell-spatial-density" class="anchor"></a>Plot flowcell spatial density</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html">include_graphics</a></span><span class="op">(</span>
  <span class="va">seqsum</span><span class="op">$</span><span class="va">flowcell</span><span class="op">$</span><span class="va">density_data</span><span class="op">$</span><span class="fu">to_file</span><span class="op">(</span><span class="st">"figure_1.png"</span><span class="op">)</span><span class="op">$</span><span class="fu">plot</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; saving plot as [png]</span></code></pre></div>
<p><img src="figure_1.png" width="100%"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stephen Rudd.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
