<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>blast_parsing • floundeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="blast_parsing">
<meta property="og:description" content="floundeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">floundeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/blast_files.html">blast_parsing</a>
    </li>
    <li>
      <a href="../articles/fast5_files.html">03 - Fast5 format sequence files</a>
    </li>
    <li>
      <a href="../articles/fasta_files.html">01 - Fasta format sequence files</a>
    </li>
    <li>
      <a href="../articles/fastq_files.html">02 - Fastq format sequence files</a>
    </li>
    <li>
      <a href="../articles/seqsum_files.html">04 - the Nanopore `Sequencing Summary` file</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="blast_files_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>blast_parsing</h1>
            
      
      
      <div class="hidden name"><code>blast_files.Rmd</code></div>

    </div>

    
    
<div id="parsing-blast-results-why" class="section level2">
<h2 class="hasAnchor">
<a href="#parsing-blast-results-why" class="anchor"></a>Parsing BLAST results? Why?</h2>
<p>NCBI BLAST (just BLAST for the rest of this document) can produce a variety of different outputs. These include the traditional and human-readable <code>Pairwise</code> format and a variety of machine readable formats that include JSON, XML and Tabular formats. The <a href="https://github.com/sagrudd/gwangmyeongseong3">Lodestar</a> project aims to enable comparative genomics and we thus require a format that can be assessed by a curious user and can be computed by software. The simplest approach would be to run each BLAST analysis twice; once for the human user and once for the machine… This doesn’t seem very environmentally friendly though.</p>
<p>Many bioinformaticians have joked that you’re not a bioinformatician until you have written yet another BLAST parser. In an attempt to expedite the rollout of the <code>floundeR</code> package (and the Lodestar functionality) a review of available R BLAST parsers was performed. There are certainly <a href="https://github.com/gschofl/blastr/">examples</a> but they are based on the more tabular output and do not seem dreadfully contemporary.</p>
<p>So, having written (much, much earlier in my career) BLAST parsers (that are lost to time) in Perl, Python and Java it is now time to again write a parser but this time in R.</p>
</div>
<div id="an-example-blast-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#an-example-blast-dataset" class="anchor"></a>An example BLAST dataset</h2>
<p>The <code>UniRef100</code> database was downloaded from <code>uniprot</code>; NCBI BLAST was installed in a <code>conda</code> environment and a BLAST database was created.</p>
<pre><code>wget ftp://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.fasta.gz
mamba install -c bioconda blast
# create the table of taxid mappings
gunzip -c uniref100.fasta.gz | \
    grep '^&gt;' | sed -e 's/&gt;//g' | sed -e 's/\s.*TaxID=/?/g' | \
    sed -e 's/\s.*//g' | sed -e 's/\?/\t/g' | grep -v 'N/A' &gt; tax_id_table
# and create the BLAST index
gunzip -c uniref100.fasta.gz | \
    makeblastdb -dbtype prot -input_type fasta -title "UniRef100" \
    -parse_seqids -hash_index -out UniRef100 -taxid_map tax_id_table
    
Building a new DB, current time: 01/11/2021 11:54:37
New DB name:   /data/UniRef100/UniRef100
New DB title:  UniRef100
Sequence type: Protein
Keep MBits: T
Maximum file size: 1000000000B

</code></pre>
<p>1000 randomly picked Drosophila sequences from a full-length cDNA analysis were compared to the <code>uniref100</code> resource using BLASTX with the standard <code>Pairwise</code> output. This sequence resource has been used for development and is reflected within the accompanying unit tests.</p>
<pre><code>mamba install -c bioconda seqkit
seqkit sample -n 250 cluster_cons.fq | seqkit fq2fa &gt; drosophila_fl_sample.fasta
[INFO] sample by number
[INFO] loading all sequences into memory...
[INFO] 248 sequences outputted</code></pre>
<p>[<a href="mailto:stephen@dellr720-16x" class="email">stephen@dellr720-16x</a> data]$ time blastx -query drosophila_fl_sample.fasta -db ./UniRef100/UniRef100 -num_threads 40 -out drosophila_fl_sample.uniref100.blastx</p>
<p>real 370m57.073s user 14621m30.880s sys 0m28.504s</p>
<p>time blastx -query drosophila_fl_sample.fasta -db ./UniRef100/UniRef100 -num_threads 40 -outfmt 18 -out drosophila_fl_sample.uniref100.blastx.org_report</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">floundeR</span><span class="op">)</span>
<span class="co">#&gt; floundeR v0.0.4</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stephen Rudd.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
