<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>R6 Class for loading, visualising and analysing barcode information — MultiplexSet • floundeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="R6 Class for loading, visualising and analysing barcode information — MultiplexSet" />
<meta property="og:description" content="R6 Class for loading, visualising and analysing barcode information
R6 Class for loading, visualising and analysing barcode information" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">floundeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.3.9003</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/fast5_review.html">fast5_review</a>
    </li>
    <li>
      <a href="../articles/fastq_processing.html">fastq_processing</a>
    </li>
    <li>
      <a href="../articles/sequencing_summary_review.html">Using the Nanopore `sequencing_summary` file for BasicQC</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R6 Class for loading, visualising and analysing barcode information</h1>
    
    <div class="hidden name"><code>MultiplexSet.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>R6 Class for loading, visualising and analysing barcode information</p>
<p>R6 Class for loading, visualising and analysing barcode information</p>
    </div>



    <h2 class="hasAnchor" id="super-class"><a class="anchor" href="#super-class"></a>Super class</h2>

    <p><code><a href='FloundeR.html'>floundeR::FloundeR</a></code> -&gt; <code>MultiplexSet</code></p>
    <h2 class="hasAnchor" id="active-bindings"><a class="anchor" href="#active-bindings"></a>Active bindings</h2>

    <p><div class="r6-active-bindings"></p><dl>
<dt><code>enumerate</code></dt><dd><p>prepares a simple <code>2D Angenieux</code> enumeration of the provided dataset
for quick visualisation of the dataset.</p></dd>

</dl><p></div></p>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>MultiplexSet$new()</code></a></p></li>
<li><p><a href='#method-as_tibble'><code>MultiplexSet$as_tibble()</code></a></p></li>
<li><p><a href='#method-read_length_bins'><code>MultiplexSet$read_length_bins()</code></a></p></li>
<li><p><a href='#method-quality_bins'><code>MultiplexSet$quality_bins()</code></a></p></li>
<li><p><a href='#method-sequencingset'><code>MultiplexSet$sequencingset()</code></a></p></li>
<li><p><a href='#method-temporalset'><code>MultiplexSet$temporalset()</code></a></p></li>
<li><p><a href='#method-clone'><code>MultiplexSet$clone()</code></a></p></li>
</ul>
<p><details open ><summary>Inherited methods</summary>
<ul>
<li><p><span class="pkg-link" data-pkg="floundeR" data-topic="FloundeR" data-id="bin_data"><a href='../../floundeR/html/FloundeR.html#method-bin_data'><code>floundeR::FloundeR$bin_data()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="floundeR" data-topic="FloundeR" data-id="num_scale"><a href='../../floundeR/html/FloundeR.html#method-num_scale'><code>floundeR::FloundeR$num_scale()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="floundeR" data-topic="FloundeR" data-id="nums_scale"><a href='../../floundeR/html/FloundeR.html#method-nums_scale'><code>floundeR::FloundeR$nums_scale()</code></a></span></p></li>
<li><p><span class="pkg-link" data-pkg="floundeR" data-topic="FloundeR" data-id="print"><a href='../../floundeR/html/FloundeR.html#method-print'><code>floundeR::FloundeR$print()</code></a></span></p></li>
</ul>
</details>

<hr>
<a id="method-new"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>new()</code></h3>
<p>Initialise a new instance of the R6 Class <code>MultiplexSet</code></p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>seqsum <span class='op'>=</span> <span class='cn'>NA</span>, barcoding_summary_file <span class='op'>=</span> <span class='cn'>NA</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>seqsum</code></dt><dd><p>a tibble of sequencing summary information</p></dd>

<dt><code>barcoding_summary_file</code></dt><dd><p>is a file.path to the corresponding
barcoding_summary file that should be merged with the <code>seqsum</code>
content.</p></dd>

</dl><p></div></p>
<p><hr>
<a id="method-as_tibble"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>as_tibble()</code></h3>
<p>Export the imported dataset(s) as a tibble</p>
<p>This object consumes a sequencing summary file (and optionally the
corresponding barcoding_summary file) and creates an object in
memory that can be explored, sliced and filtered. This method dumps
out the in-memory object for further exploration and development.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>as_tibble</span><span class='op'>(</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p>A tibble representation of the starting dataset</p>
<p><hr>
<a id="method-read_length_bins"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>read_length_bins()</code></h3>
<p>bin the sequences in <code>seqsum</code> content into bins of sequence length</p>
<p>The nanopore sequencing run is expected to return a collection of
sequences that vary in their length distributions; this variance is a
function of the sequencing library prepared, the starting DNA etc. This
method is used to bin reads into uniform bins &amp; assess the distribution
of sequence lengths.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>read_length_bins</span><span class='op'>(</span>
  qfilt <span class='op'>=</span> <span class='cn'>TRUE</span>,
  normalised <span class='op'>=</span> <span class='cn'>TRUE</span>,
  cumulative <span class='op'>=</span> <span class='cn'>FALSE</span>,
  bins <span class='op'>=</span> <span class='fl'>20</span>,
  outliers <span class='op'>=</span> <span class='fl'>0.025</span>
<span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>qfilt</code></dt><dd><p><ul>
<li><p>specifies how the quality information should be filtered</p></li>
<li><p>at the moment this only defines whether reporting is based on PASS
or FAIL reads; would make more sense to have filtered by PASS / ALL?</p></li>
</ul></p></dd>

<dt><code>normalised</code></dt><dd><p><ul>
<li><p>should the sequence collection be reported to
normalise for the number of sequence bases sequenced or the number of
sequence reads - TRUE by default to normalise for sequenced bases.</p></li>
</ul></p></dd>

<dt><code>cumulative</code></dt><dd><p>defines whether cumulative sequence bases (reads) are
reported per bin (FALSE by default).</p></dd>

<dt><code>bins</code></dt><dd><p>the number of sequence bins that should be prepared (20 by
default)</p></dd>

<dt><code>outliers</code></dt><dd><p>defines the number of outliers (0.025 = 2.5%) that are
excluded from the longest reads to prepare a richer distribution
visulation - the plots can be bothered by the long tail of mini-whales.</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p>Angenieux 2D graph object</p>
<p><hr>
<a id="method-quality_bins"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>quality_bins()</code></h3>
<p>bin the sequences in <code>seqsum</code> content into bins of quality</p>
<p>The nanopore sequencing run is expected to return a collection of
sequences that vary in their quality distributions; this variance is a
function of the sequencing library prepared, the starting DNA etc. This
method is used to bin reads into uniform quality bins to assess the
overall quality of the run and to identify potential issues</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>quality_bins</span><span class='op'>(</span>bins <span class='op'>=</span> <span class='fl'>20</span>, outliers <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>bins</code></dt><dd><p>the number of sequence bins that should be prepared (20 by
default)</p></dd>

<dt><code>outliers</code></dt><dd><p>defines the number of outliers (0 = 0%) that are
excluded from the reads - should probably be deprecated for simplicity?</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p>Angenieux 2D graph object</p>
<p><hr>
<a id="method-sequencingset"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>sequencingset()</code></h3>
<p>Prepare a SequencingSet object from a given barcode</p>
<p>This method is used to subset the sequencing_summary information to
focus on a single barcode for more detailed analysis.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>sequencingset</span><span class='op'>(</span><span class='va'>barcode</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>barcode</code></dt><dd><p>the barcode that should be reported</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p><code>SequencingSet</code> object containing information on barcode of
interest.</p>
<p><hr>
<a id="method-temporalset"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>temporalset()</code></h3>
<p>Prepare a TemporalSet object from a given barcode</p>
<p>This method is used to subset the SequencingSummary information to
focus on a single barcode for a more detailed analysis of content.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>temporalset</span><span class='op'>(</span><span class='va'>barcode</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>barcode</code></dt><dd><p>the barcode that should be reported</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p><code>TemporalSet</code> object containing information on barcode of
interest.</p>
<p><hr>
<a id="method-clone"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>clone()</code></h3>
<p>The objects of this class are cloneable with this method.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>MultiplexSet</span><span class='op'>$</span><span class='fu'>clone</span><span class='op'>(</span>deep <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>deep</code></dt><dd><p>Whether to make a deep clone.</p></dd>

</dl><p></div></p>



  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Stephen Rudd.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


